Recursive dependency involving fixture 'celery_config' if Flask-CeleryExt is not installed

**Disclaimer**: I might be the source of this issue https://github.com/inveniosoftware/pytest-invenio/commit/fc7ee039a8995be851df1df7bcc560ea4c87b226#diff-2eeaed663bd0d25b7e608891384b7298L47

When [Flask-CeleryExt](https://github.com/inveniosoftware/flask-celeryext) is not installed by the module where `pytest-invenio` is used, this error appears:

```
file .../src/inveniosoftware/invenio-sequencegenerator/tests/test_models.py, line 167
  def test_exceptions(db):
file .../lib/python3.5/site-packages/pytest_invenio/fixtures.py, line 353
  @pytest.fixture(scope='module')
  def database(appctx):
file .../lib/python3.5/site-packages/pytest_invenio/fixtures.py, line 256
  @pytest.fixture(scope='module')
  def appctx(base_app):
file .../lib/python3.5/site-packages/pytest_invenio/fixtures.py, line 177
  @pytest.fixture(scope='module')
  def base_app(create_app, app_config, request, default_handler):
file .../lib/python3.5/site-packages/pytest_invenio/fixtures.py, line 124
  @pytest.fixture(scope='module')
  def app_config(db_uri, broker_uri, celery_config):
file .../lib/python3.5/site-packages/pytest_invenio/fixtures.py, line 95
  @pytest.fixture(scope='module')
  def celery_config(celery_config):
E       recursive dependency involving fixture 'celery_config' detected
>       available fixtures: _configure_application, _monkeypatch_response_class, _push_request_context, accept_any, accept_json, accept_jsonp, accept_mimetype, app, app_config, appctx, base_app, base_client, broker_uri, browser, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, celery_config, cli_runner, client, client_class, config, cov, create_app, database, db, db_uri, default_handler, doctest_namespace, es, es_clear, instance_path, live_server, mailbox, monkeypatch, no_cover, pytestconfig, record_property, record_xml_attribute, record_xml_property, recwarn, request_ctx, script_info, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory
>       use 'pytest --fixtures [testpath]' for help on them.
```

This happens basically because `celery_config` fixture doesn't exists and therefore references itself [here](https://github.com/inveniosoftware/pytest-invenio/blob/master/pytest_invenio/fixtures.py#L96).

One solution might be to add back `flask-celeryext` to the requirements, not only the test requirements, but this would make modules, which don't need celery, install it for the tests.

From the top of my head I can't find an elegant solution, perhaps someone better versed on pytest might know something I don't. All I could come up is something along this lines:

```
try:
    pkg_resources.get_distribution('flask-celeryext')
    @pytest.fixture(scope='module')
    def celery_config(celery_config):
        .....
except pkg_resources.DistributionNotFound:
    @pytest.fixture(scope='module')
    def celery_config():
        ....
```

--- a/pytest_invenio/fixtures.py
+++ b/pytest_invenio/fixtures.py
@@ -17,6 +17,7 @@ import shutil
 import tempfile
 from datetime import datetime
 
+import pkg_resources
 import pytest
 from pytest_flask.plugin import _make_test_response_class
 from selenium import webdriver
@@ -92,33 +93,59 @@ def broker_uri():
     yield os.environ.get('BROKER_URL', 'amqp://guest:guest@localhost:5672//')
 
 
-@pytest.fixture(scope='module')
-def celery_config(celery_config):
-    """Celery configuration (defaults to eager tasks).
+def _celery_config():
+    """Factory/Helper function to create the ``celery_config`` fixture.
 
-    Scope: module
-
-    This fixture provides the default Celery configuration (eager tasks,
-    in-memory result backend and exception propagation). It can easily be
-    overwritten in a specific test module:
-
-    .. code-block:: python
-
-        # test_something.py
-        import pytest
-
-        pytest.fixture(scope='module')
-        def celery_config(celery_config):
-            celery_config['CELERY_ALWAYS_EAGER'] = False
-            return celery_config
+    When Celery is installed it provides with this same fixture via
+    `celery.contrib.pytest
+    <https://github.com/celery/celery/blob/master/celery/contrib/pytest.py>`_,
+    in this is the case we overwrite this fixture and update the configuration
+    with Invenio's default configuration.
+    If it is not installed, then we just define a new fixture which returns the
+    default Invenio Celery configuration.
     """
-    celery_config.update(dict(
+    default_config = dict(
         CELERY_ALWAYS_EAGER=True,
         CELERY_CACHE_BACKEND='memory',
         CELERY_EAGER_PROPAGATES_EXCEPTIONS=True,
         CELERY_RESULT_BACKEND='cache',
-    ))
-    return celery_config
+    )
+
+    try:
+        pkg_resources.get_distribution('celery')
+
+        # Celery is installed, overwrite fixture
+        def inner(celery_config):
+            celery_config.update(default_config)
+            return celery_config
+    except pkg_resources.DistributionNotFound:
+        # No Celery, return the default config
+        def inner():
+            return default_config
+
+    return inner
+
+
+celery_config = pytest.fixture(
+    scope='module', name='celery_config')(_celery_config())
+"""Celery configuration (defaults to eager tasks).
+
+Scope: module
+
+This fixture provides the default Celery configuration (eager tasks,
+in-memory result backend and exception propagation). It can easily be
+overwritten in a specific test module:
+
+.. code-block:: python
+
+    # test_something.py
+    import pytest
+
+    pytest.fixture(scope='module')
+    def celery_config(celery_config):
+        celery_config['CELERY_ALWAYS_EAGER'] = False
+        return celery_config
+"""
 
 
 @pytest.fixture(scope='module')

--- a/tests/test_fixtures.py
+++ b/tests/test_fixtures.py
@@ -377,3 +377,15 @@ def test_browser(conftest_testdir, monkeypatch):
     assert os.path.exists(
         os.path.join(str(conftest_testdir.tmpdir), '.e2e_screenshots'))
     monkeypatch.undo()
+
+
+def test_celery_config(testdir):
+    """Test celery config."""
+    testdir.makepyfile(test_app="""
+        def test_celery_config_with_celery(celery_config):
+            assert celery_config['CELERY_ALWAYS_EAGER'] == True
+            assert celery_config['CELERY_CACHE_BACKEND'] == 'memory'
+            assert celery_config['CELERY_EAGER_PROPAGATES_EXCEPTIONS'] == True
+            assert celery_config['CELERY_RESULT_BACKEND'] == 'cache'
+    """)
+    testdir.runpytest().assert_outcomes(passed=1)

